tests/compute/test_heterograph.py

```diff
@@ -10,151 +9,388 @@
+        'plays': ([0, 2], [1, 0]),
+        'wishes': ([0, 1], [0, 1]),
+        'develops': ([0, 1], [1, 0]),
+    }
+    _test_query()
+    etypes = canonical_etypes
+    edges = {
+        ('user', 'follows', 'user'): ([0, 1], [1, 2]),
+        ('user', 'plays', 'game'): ([0, 1, 2, 1], [0, 0, 1, 1]),
+        ('user', 'wishes', 'game'): ([0, 2], [1, 0]),
+        ('developer', 'develops', 'game'): ([0, 1], [0, 1]),
+    }
+    # edges that does not exist in the graph
+    negative_edges = {
+        ('user', 'follows', 'user'): ([0, 1], [0, 1]),
+        ('user', 'plays', 'game'): ([0, 2], [1, 0]),
+        ('user', 'wishes', 'game'): ([0, 1], [0, 1]),
+        ('developer', 'develops', 'game'): ([0, 1], [1, 0]),
+        }
+    _test_query()
+
+    # test features
+    HG.nodes['user'].data['h'] = F.ones((HG.number_of_nodes('user'), 5))
+    HG.nodes['game'].data['m'] = F.ones((HG.number_of_nodes('game'), 3)) * 2
+
+    # test only one node type
+    g = HG['follows']
+    assert g.number_of_nodes() == 3
+
+    # test ndata and edata
+    f1 = F.randn((3, 6))
+    g.ndata['h'] = f1       # ok
+    f2 = HG.nodes['user'].data['h']
     assert F.array_equal(f1, f2)
-    assert F.array_equal(g.nodes['user'][0].data['h'], f1[0:1])
+    assert F.array_equal(g.nodes(), F.arange(0, 3))
 
     f3 = F.randn((2, 4))
-    g.edata['user', 'follows', 'user']['h'] = f3
-    f4 = g.edata['user', 'follows', 'user']['h']
+    g.edata['h'] = f3
+    f4 = HG.edges['follows'].data['h']
     assert F.array_equal(f3, f4)
-    assert F.array_equal(g.edges['user', 'follows', 'user'][0].data['h'], f3[0:1])
+    assert F.array_equal(g.edges(form='eid'), F.arange(0, 2))
+
+    # test fail case
+    # fail due to multiple types
+    fail = False
+    try:
+        HG.ndata['h']
+    except dgl.DGLError:
+        fail = True
+    assert fail
+
+    fail = False
+    try:
+        HG.edata['h']
+    except dgl.DGLError:
+        fail = True
+    assert fail
+
+def test_flatten():
+    def check_mapping(g, fg):
+        if len(fg.ntypes) == 1:
+            SRC = DST = dgl.DEFAULT
+        else:
+            SRC = dgl.SRC
+            DST = dgl.DST
+
+        etypes = F.asnumpy(fg.edata[dgl.ETYPE]).tolist()
+        eids = F.asnumpy(fg.edata[dgl.EID]).tolist()
+
+        for i, (etype, eid) in enumerate(zip(etypes, eids)):
+            src_g, dst_g = g.find_edges([eid], g.canonical_etypes[etype])
+            src_fg, dst_fg = fg.find_edges([i])
+            # TODO(gq): I feel this code is quite redundant; can we just add new members (like
+            # "induced_srcid") to returned heterograph object and not store them as features?
+            assert src_g == fg.nodes[SRC].data[dgl.NID][src_fg]
+            assert g.canonical_etypes[etype][0] == g.ntypes[fg.nodes[SRC].data[dgl.NTYPE][src_fg]]
+            assert dst_g == fg.nodes[DST].data[dgl.NID][dst_fg]
+            assert g.canonical_etypes[etype][2] == g.ntypes[fg.nodes[DST].data[dgl.NTYPE][dst_fg]]
```

Comment:
I feel it's pretty ok if you use a variable here:
```python
mapping = fg.nodes[SRC].data[dgl.NID]
assert src_g = mapping[src_fg]
```
To make it better, I think we should try to shorten the `fg.nodes[type].data[fld]`.

Reply:
No, I mean, I feel the code would look cleaner if the mapping can be accessed with an attribute like `fg.induced_srcid`.

Maybe we could provide a property that acts as a shortcut to the reserved column.

```python
@property
def induced_srcid(self):
    return self.nodes[SRC].data[dgl.NID]
```

Reply:
Comparing the two:
```python
mapping = fg.nodes[SRC].data[dgl.NID]
assert src_g = mapping[src_fg]
```
and
```python
mapping = fg.induced_srcid
assert src_g = mapping[src_fg]
```
I feel the difference is not that huge and the learning effort is similar, if we could beautify the `fg.nodes[SRC].data[dgl.NID]` syntax (for example, `fg.ndata['id', 'src']`). Another option is to have `g.srcdata` and `g.dstdata` and add `mapid` to the data view. This ends up with:
```python
mapping = fg.srcdata.mapid
```
Another question, if the slice returns a homograph, will `induced_srcid` still make sense?